server/src/sas/Lexer.ts

```diff
@@ -235,37 +244,120 @@ export class Lexer {
+    }
+    // proc python/sql/lua...
+    switch (this.context.embeddedLangState) {
+      case EmbeddedLangState.NONE: {
+        if (
+          this.context.lastNoncommentToken?.type === "text" &&
+          this.context.lastNoncommentToken.text === "PROC"
+        ) {
+          if (
+            token.type === "text" &&
+            (token.text === "SQL" || token.text === "FEDSQL")
+          ) {
+            this.context.embeddedLangState = EmbeddedLangState.PROC_SQL_DEF;
+          } else if (token.type === "text" && token.text === "PYTHON") {
+            this.context.embeddedLangState = EmbeddedLangState.PROC_PYTHON_DEF;
+          } else if (token.type === "text" && token.text === "LUA") {
+            this.context.embeddedLangState = EmbeddedLangState.PROC_LUA_DEF;
+          }
+        }
+        break;
+      }
+      case EmbeddedLangState.PROC_PYTHON_DEF: {
+        if (
+          token.type === "text" &&
+          ["SUBMIT", "INTERACTIVE"].includes(token.text)
+        ) {
+          this.context.embeddedLangState =
+            EmbeddedLangState.PROC_PYTHON_SUBMIT_OR_INTERACTIVE;
+        }
+        break;
       }
-      if (Lexer.isLiteral[token.type]) {
-        token.text = this.getText(token);
-      } else {
-        token.text = this.getWord(token).toUpperCase();
+      case EmbeddedLangState.PROC_LUA_DEF: {
+        if (
+          token.type === "text" &&
+          ["SUBMIT", "INTERACTIVE"].includes(token.text)
+        ) {
+          this.context.embeddedLangState =
+            EmbeddedLangState.PROC_LUA_SUBMIT_OR_INTERACTIVE;
+        }
+        break;
       }
-      if (Lexer.isComment[token.type] === undefined) {
-        this.quoting = checkQuote(
-          this.quoting,
-          !!Lexer.isQuoting[token.text],
-          token.text,
-        );
-        this.bquoting = checkQuote(
-          this.bquoting,
-          !!Lexer.isBQuoting[token.text],
-          token.text,
-        );
-        if (this.quoting === -1 && this.bquoting === -1) {
-          if (!this.ignoreFormat && token.text === "%PUT") {
-            this.ignoreFormat = true;
-          } else if (this.ignoreFormat && token.text === ";") {
-            this.ignoreFormat = false;
+      case EmbeddedLangState.PROC_PYTHON_SUBMIT_OR_INTERACTIVE:
+      case EmbeddedLangState.PROC_LUA_SUBMIT_OR_INTERACTIVE:
+      case EmbeddedLangState.PROC_SQL_DEF: {
+        if (token.type === "sep" && token.text === ";") {
+          let endReg;
+          if (
+            this.context.embeddedLangState ===
+              EmbeddedLangState.PROC_PYTHON_SUBMIT_OR_INTERACTIVE ||
+            this.context.embeddedLangState ===
+              EmbeddedLangState.PROC_LUA_SUBMIT_OR_INTERACTIVE
+          ) {
+            endReg =
+              /\b(endsubmit|endinteractive|quit|run)(\s+|\/\*.*?\*\/)*;(\s+|\/\*.*?\*\/)*$/i;
+          } else {
+            endReg = /\b(run|quit)(\s+|\/\*.*?\*\/)*;(\s+|\/\*.*?\*\/)*$/i;
```

Comment:
proc sql will only end by `quit` not `run`

Another thing is that, proc can end without explicit quit, when another proc met, refers to sample code
https://go.documentation.sas.com/doc/en/pgmsascdc/v_048/sqlproc/n1ncn0pznd8wrln1tnp3xdxjz9xz.htm

Reply:
Seems it can end by run:  https://go.documentation.sas.com/doc/en/pgmsascdc/v_048/proc/p0o588yrrlokr3n1osm419b7hdtg.htm

Reply:
If you try run
```
proc sql;
    select * from sashelp.cars;
run;
    select * from sashelp.cars;
```
You'll find that `run` will not end sql proc; The 2nd select statement still works. You'll got below note from SAS log:
`NOTE: PROC SQL statements are executed immediately; The RUN statement has no effect.`